controller/sync.go

```diff
@@ -354,29 +356,57 @@ func (sc *syncContext) pruneObject(liveObj *unstructured.Unstructured, prune, dr
 // Or if the prune/apply failed, will also update the result.
 func (sc *syncContext) doApplySync(syncTasks []syncTask, dryRun, force, update bool) bool {
 	syncSuccessful := true
-	// apply all resources in parallel
+
+	createTasks := []syncTask{}
+	pruneTasks := []syncTask{}
+	for _, syncTask := range syncTasks {
+		if syncTask.targetObj == nil {
+			pruneTasks = append(pruneTasks, syncTask)
+		} else {
+			createTasks = append(createTasks, syncTask)
+
+		}
+	}
+
 	var wg sync.WaitGroup
-	for _, task := range syncTasks {
+	for _, task := range pruneTasks {
 		wg.Add(1)
 		go func(t syncTask) {
 			defer wg.Done()
 			var resDetails appv1.ResourceDetails
-			if t.targetObj == nil {
-				resDetails = sc.pruneObject(t.liveObj, sc.syncOp.Prune, dryRun)
-			} else {
-				if isHook(t.targetObj) {
-					return
-				}
-				resDetails = sc.applyObject(t.targetObj, dryRun, force)
+			resDetails = sc.pruneObject(t.liveObj, sc.syncOp.Prune, dryRun)
+			if !resDetails.Status.Successful() {
+				syncSuccessful = false
+			}
+			if update || !resDetails.Status.Successful() {
+				sc.setResourceDetails(&resDetails)
+			}
+		}(task)
+	}
+	currKind := ""
+	var createWg sync.WaitGroup
```

Comment:
Looks like sync tasks are sorted in the correct order because application resources are sorted in `AppStateManager`. This looks fragile and could be broken. 

Instead of applying sorting in `AppStateManager` please sort sync tasks in `sync` method.